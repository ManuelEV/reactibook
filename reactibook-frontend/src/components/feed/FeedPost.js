import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import 'moment/locale/es';
import { startDeleting, startUpdatingPost } from '../../actions/posts';

export const FeedPost = ({id, content, date, fileUrl, author}) => {

    moment().locale('es');

    const postDate = moment(date);

    const { uid } = useSelector( state => state.auth );

    const dispatch = useDispatch();

    const [editionMode, setEditionMode] = useState(false);

    const [editedPostContent, setEditedPostContent] = useState(content);

    const handleInputChange = (e) => {
        setEditedPostContent(e.target.value);
    }

    const [formErrorMsg, setformErrorMsg] = useState('');

    const isFormValid = () => {

        if ( content === editedPostContent ) {
            setformErrorMsg('El contenido del post no debe ser el mismo que el anterior');
            return false;
        } else if ( editedPostContent.trim().length === 0 ) {
            setformErrorMsg('El contenido del post no debe estar vacío');
            return false;
        }          

        return true;
    }

    const handlePostUpdate = () => {
        if (isFormValid()) {
            dispatch(startUpdatingPost(id, {content: editedPostContent}));
            setEditionMode(false);
        }
    }

    const handlePostDelete = () => {
        dispatch(startDeleting(id));
    }

    return (
        <div className="w-full flex justify-center">
            <div className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
                rounded">
                <div 
                    className="flex row mb-2"
                >
                    <img 
                            className="h-14 w-14"
                            src="./assets/user.svg" alt="user_profile"
                    />
                    <div className="grid grid-cols-1 ml-3">
                        <div className="text-xl font-semibold">{author.name}</div>
                        <div className="text-sm">{postDate.calendar()}</div>
                    </div>
                </div>

                
                {
                    editionMode ? (
                        <div className="w-full">
                            <textarea 
                                className="w-full rounded flex items-start p-2 text-black resize-none"
                                type="textarea"
                                autoComplete="off"
                                placeholder="¿Qué está pasando?"
                                name="editedPostContent"
                                value={editedPostContent}
                                onChange={handleInputChange}
                            >
                            </textarea>
                            <span className="text-red-500 text-sm">
                                {formErrorMsg}
                            </span>
                            <div className="w-full flex justify-end px-3 text-sm underline">
                                <span 
                                    className="pr-2 cursor-pointer"
                                    onClick={() => setEditionMode(false)}
                                >
                                    Cancelar
                                </span>
                                <span
                                    className="cursor-pointer"
                                    onClick={handlePostUpdate}
                                >
                                    Guardar
                                </span>
                            </div>
                        </div>
                    )

                    : (
                        <div 
                            className="w-full rounded flex items-start p-2"
                        >
                            {content}
                        </div>
                    )
                }

                <div 
                    className="w-full flex justify-center"
                >
                    <img src={fileUrl} alt="" />
                </div>
                
                {
                    (uid === author.id) &&
                    (<div className="w-full flex items-start px-3 text-sm underline">
                        <span 
                            className="pr-2 cursor-pointer"
                            onClick={() => setEditionMode(true)}
                        >
                            Editar
                        </span>
                        <span
                            className="cursor-pointer"
                            onClick={handlePostDelete}
                        >
                            Eliminar
                        </span>
                    </div>)
                }
            </div>
        </div>
    )
}
