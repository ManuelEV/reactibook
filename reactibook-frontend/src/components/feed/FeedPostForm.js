import React from 'react'

export const FeedPostForm = () => {
    return (
        <div className="w-full flex justify-center">
            <form className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
                rounded">
                <textarea 
                    className="w-full rounded flex items-start p-2 text-black resize-none"
                    type="textarea"
                    autoComplete="off"
                    placeholder="¿Qué está pasando?"
                    
                ></textarea>

                <div className="w-full flex row mt-2 mx-2">
                    <div className="w-1/3 flex items-start">
                        <input
                            className=""
                            type="file"
                            name="file"
                            accept="image/png, image/jpeg"
                        />
                    </div>
                    <div className="w-2/3 flex justify-end">
                        <select
                            className="mr-2 text-black"
                            name="postVisibility"
                        >
                            <option value="friends">Amigos</option>
                            <option value="public">Público</option>
                        </select>
                        <button
                            className="bg-primary-600 rounded px-2 text-white
                            hover:bg-primary-500"
                        >
                            publicar
                        </button>
                    </div>
                </div>
            </form>
        </div>
    )
}
