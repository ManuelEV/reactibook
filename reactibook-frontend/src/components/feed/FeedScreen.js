import React, { useState } from 'react'
import { useSelector } from 'react-redux';
import { NavBar } from '../navigation/NavBar'
import { FeedPost } from './FeedPost';
import { FeedPostForm } from './FeedPostForm'

export const FeedScreen = () => {

    const {posts} = useSelector( state => state.posts );

    const [filteredPosts, setPosts] = useState(posts);

    const handleFilterPosts = (filterCondition) => {
        setPosts( posts.filter(post => post.visibility === filterCondition) );
    }

    return (
        <div className="w-full h-full">

            <NavBar />
            
            <div className="pt-16">

                <FeedPostForm />

                <div className="w-full flex justify-center">
                    <div className="w-3/5 flex row">
                        <span 
                            className="px-2 underline cursor-pointer"
                            onClick={() => handleFilterPosts('public')}
                        >
                            PÃºblico
                        </span>
                        <span 
                            className="px-2 underline cursor-pointer"
                            onClick={() => handleFilterPosts('friends')}
                        >
                            Amigos
                        </span>
                        <span 
                            className="px-2 underline cursor-pointer"
                            onClick={() => setPosts(posts)}
                        >
                            Ver todo
                        </span>
                    </div>
                </div>

                {
                    filteredPosts.map(post => (
                        <FeedPost
                            key={post.id}
                            {...post}
                        />
                    ))
                }
            </div>

        </div>
    )
}
