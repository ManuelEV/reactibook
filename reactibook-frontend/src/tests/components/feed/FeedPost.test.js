import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { FeedPost } from '../../../components/feed/FeedPost';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {uid: 'id', name: 'test'},
    ui: {
        loading: false,
        msgError: null,
        loginErrors: {
            emailMsg: '',
            passwordMsg: '',
        },
    },
    posts: {
        posts: [],
        currentFileUrl: ''
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const post = {
    id: 'id',
    author: {id: 'id', name: 'name'},
    content: 'content',
    date: 10000000,
    fileUrl: '',
};

const wrapper = mount( 
    <Provider store={ store }>
        <FeedPost { ...post }  /> 
    </Provider>
)


describe('Tests in <FeedPost/> component', () => {
    

    test('snapshot should match', () => {
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('the textarea to edit the post should not exist', () => {

        expect(wrapper.find('textarea').exists()).toBe(false);

    });

    test('if the id of the user matches the author of the post, you should be able to edit and delete', () => {

        const postSpans = wrapper.find('span');

        expect(postSpans.at(0).text()).toBe('Editar');

        expect(postSpans.at(1).text()).toBe('Eliminar');

    });

})
