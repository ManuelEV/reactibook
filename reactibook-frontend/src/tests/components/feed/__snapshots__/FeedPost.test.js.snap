// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <FeedPost/> component snapshot should match 1`] = `
<div
  className="w-full flex justify-center"
>
  <div
    className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
  >
    <div
      className="flex row mb-2"
    >
      <img
        alt="user_profile"
        className="h-14 w-14"
        src="./assets/user.svg"
      />
      <div
        className="grid grid-cols-1 ml-3"
      >
        <div
          className="text-xl font-semibold"
        >
          name
        </div>
        <div
          className="text-sm"
        >
          31/12/1969
        </div>
      </div>
    </div>
    <div
      className="w-full rounded flex items-start p-2"
    >
      content
    </div>
    <div
      className="w-full flex justify-center"
    >
      <img
        alt=""
        src=""
      />
    </div>
    <div
      className="w-full flex items-start px-3 text-sm underline"
    >
      <span
        className="pr-2 cursor-pointer"
        onClick={[Function]}
      >
        Editar
      </span>
      <span
        className="cursor-pointer"
        onClick={[Function]}
      >
        Eliminar
      </span>
    </div>
  </div>
</div>
`;
